File: Javascript-DSA/Searching/AlgoExpert/binarySearch.js

function binarySearch(array, target) {
    return binarySearchHelper(array, target, 0, array.length - 1);
}

//! O(nlogn) time | O(1) space
function binarySearchHelper(array, target, left, right) {
    
    while(left <= right) {
        const middle = Math.floor((left + right) / 2);
        const potentialMatch = array[middle];
        if(target == potentialMatch) {
            return middle;
        } else if(target < potentialMatch) {
            right = middle - 1;
        } else {
            left = middle + 1;
        }
    }
    return -1;
}

//! O(nlogn) time | O(logn) space
function binarySearchHelper(array, target, left, right) {

    if(left > right) return -1;

    const middle = Math.floor((left + right) / 2);
    const potentialMatch = array[middle];
    if(target == potentialMatch) {
        return middle;
    } else if(target < potentialMatch) {
        return binarySearchHelper(array, target, left, middle - 1);
    } else {
        return binarySearchHelper(array, target, middle + 1, right);
    }
}

File: Searching/BinarySearch.java
package Searching;

public class BinarySearch {

    public static void main(String args[]) {
        int[] array = {0, 1, 21, 33, 45, 45, 61, 71, 72, 73};
        int target = 33;
        System.out.println(binarySearchRecursion(array, target, 0, array.length - 1));
    }
    public static int binarySearch(int array[], int target, int left, int right) {
        // O(log(n)) time | O(1) space
        while(left <= right) {
            int mid = left + (right - left) / 2;
            int potentialMatch = array[mid];
            if(target == potentialMatch) return mid;
            else if(target < potentialMatch) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }
    public static int binarySearchRecursion(int array[], int target, int left, int right) {
        // O(log(n)) time | O(log(n)) space
        if(left > right) return -1;
        int mid = left + (right - left) / 2;
        int potentialMatch = array[mid];
        if(target == potentialMatch) return mid;
        else if(target < potentialMatch)
            return binarySearchRecursion(array, target, left, mid - 1);
        else
            return binarySearchRecursion(array, target, mid + 1, right);
    }
}
